### ISSUE TITLE
ConnectionError: HTTPConnectionPool(host='localhost', port=4040): Max retries exceeded with url: /api/tunnels 

### ISSUE THREAD

**Issue by koluzaeva** [SwayamInSync/clothes-virtual-try-on#22]
Hi! I try run app() but receive an error:
```
* Serving Flask app '__main__'
 * Debug mode: off
INFO:werkzeug:WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on http://127.0.0.1:5000/
INFO:werkzeug:Press CTRL+C to quit
Exception in thread Thread-13:
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/urllib3/connection.py", line 203, in _new_conn
    sock = connection.create_connection(
  File "/usr/local/lib/python3.10/dist-packages/urllib3/util/connection.py", line 85, in create_connection
    raise err
  File "/usr/local/lib/python3.10/dist-packages/urllib3/util/connection.py", line 73, in create_connection
    sock.connect(sa)
ConnectionRefusedError: [Errno 111] Connection refused

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/urllib3/connectionpool.py", line 791, in urlopen
    response = self._make_request(
  File "/usr/local/lib/python3.10/dist-packages/urllib3/connectionpool.py", line 497, in _make_request
    conn.request(
  File "/usr/local/lib/python3.10/dist-packages/urllib3/connection.py", line 395, in request
    self.endheaders()
  File "/usr/lib/python3.10/http/client.py", line 1278, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.10/http/client.py", line 1038, in _send_output
    self.send(msg)
  File "/usr/lib/python3.10/http/client.py", line 976, in send
    self.connect()
  File "/usr/local/lib/python3.10/dist-packages/urllib3/connection.py", line 243, in connect
    self.sock = self._new_conn()
  File "/usr/local/lib/python3.10/dist-packages/urllib3/connection.py", line 218, in _new_conn
    raise NewConnectionError(
urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x7dae50db8520>: Failed to establish a new connection: [Errno 111] Connection refused

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/requests/adapters.py", line 486, in send
    resp = conn.urlopen(
  File "/usr/local/lib/python3.10/dist-packages/urllib3/connectionpool.py", line 845, in urlopen
    retries = retries.increment(
  File "/usr/local/lib/python3.10/dist-packages/urllib3/util/retry.py", line 515, in increment
    raise MaxRetryError(_pool, url, reason) from reason  # type: ignore[arg-type]
urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='localhost', port=4040): Max retries exceeded with url: /api/tunnels (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7dae50db8520>: Failed to establish a new connection: [Errno 111] Connection refused'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.10/threading.py", line 1378, in run
    self.function(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.10/dist-packages/flask_ngrok.py", line 70, in start_ngrok
    ngrok_address = _run_ngrok()
  File "/usr/local/lib/python3.10/dist-packages/flask_ngrok.py", line 35, in _run_ngrok
    tunnel_url = requests.get(localhost_url).text  # Get the tunnel information
  File "/usr/local/lib/python3.10/dist-packages/requests/api.py", line 73, in get
    return request("get", url, params=params, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/requests/api.py", line 59, in request
    return session.request(method=method, url=url, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/requests/sessions.py", line 589, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python3.10/dist-packages/requests/sessions.py", line 703, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/requests/adapters.py", line 519, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPConnectionPool(host='localhost', port=4040): Max retries exceeded with url: /api/tunnels (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7dae50db8520>: Failed to establish a new connection: [Errno 111] Connection refused'))  
```
Can you please help me?

**Comment by SwayamInSync** [comment]
@koluzaeva 

It is an ngrok error, indicating that ngrok server is not running

To fix it, 
1. Disconnect the runtime
2. Run the following command in the first cell of entire colab notebook

```bash
!pip install flask-ngrok
!pip install pyngrok==4.1.1
!ngrok authtoken <NGROK_TOKEN>
```

**Comment by koluzaeva** [comment]
@SwayamInSync thank you for your answer. I changed ngrok key and now there is no error and output looks like this:
```
* Serving Flask app '__main__'
 * Debug mode: off
INFO:werkzeug:WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on http://127.0.0.1:5000/
INFO:werkzeug:Press CTRL+C to quit
 * Running on http://f045-34-126-69-210.ngrok-free.app/
 * Traffic stats available on http://127.0.0.1:4040/
INFO:werkzeug:127.0.0.1 - - [18/Dec/2023 14:35:41] "GET / HTTP/1.1" 200 -
INFO:werkzeug:127.0.0.1 - - [18/Dec/2023 14:35:42] "GET /favicon.ico HTTP/1.1" 404 -
INFO:werkzeug:127.0.0.1 - - [18/Dec/2023 14:36:12] "GET / HTTP/1.1" 200 -
INFO:werkzeug:127.0.0.1 - - [18/Dec/2023 14:36:14] "GET / HTTP/1.1" 200 -
INFO:werkzeug:127.0.0.1 - - [18/Dec/2023 14:37:03] "GET / HTTP/1.1" 200 -
INFO:werkzeug:127.0.0.1 - - [18/Dec/2023 14:37:41] "GET / HTTP/1.1" 200 -
INFO:werkzeug:127.0.0.1 - - [18/Dec/2023 14:37:51] "GET /api/transform HTTP/1.1" 405 -
INFO:werkzeug:127.0.0.1 - - [18/Dec/2023 14:38:01] "GET / HTTP/1.1" 200 -
```
but http://127.0.0.1:5000/ - "The site can't be reached". 
at http://79c9-34-126-69-210.ngrok-free.app/ I see "Hello, world!"

**Comment by SwayamInSync** [comment]
> @SwayamInSync thank you for your answer. I changed ngrok key and now there is no error and output looks like this:
> 
> ```
> * Serving Flask app '__main__'
>  * Debug mode: off
> INFO:werkzeug:WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
>  * Running on http://127.0.0.1:5000/
> INFO:werkzeug:Press CTRL+C to quit
>  * Running on http://f045-34-126-69-210.ngrok-free.app/
>  * Traffic stats available on http://127.0.0.1:4040/
> INFO:werkzeug:127.0.0.1 - - [18/Dec/2023 14:35:41] "GET / HTTP/1.1" 200 -
> INFO:werkzeug:127.0.0.1 - - [18/Dec/2023 14:35:42] "GET /favicon.ico HTTP/1.1" 404 -
> INFO:werkzeug:127.0.0.1 - - [18/Dec/2023 14:36:12] "GET / HTTP/1.1" 200 -
> INFO:werkzeug:127.0.0.1 - - [18/Dec/2023 14:36:14] "GET / HTTP/1.1" 200 -
> INFO:werkzeug:127.0.0.1 - - [18/Dec/2023 14:37:03] "GET / HTTP/1.1" 200 -
> INFO:werkzeug:127.0.0.1 - - [18/Dec/2023 14:37:41] "GET / HTTP/1.1" 200 -
> INFO:werkzeug:127.0.0.1 - - [18/Dec/2023 14:37:51] "GET /api/transform HTTP/1.1" 405 -
> INFO:werkzeug:127.0.0.1 - - [18/Dec/2023 14:38:01] "GET / HTTP/1.1" 200 -
> ```
> 
> but http://127.0.0.1:5000/ - "The site can't be reached". at http://79c9-34-126-69-210.ngrok-free.app/ I see "Hello, world!"

This is expected, now use the public URL (ending with .ngrok-free.app) to make the post request to `public-url/api/transform` from the client side (provided in the repo, or you can use the postman for same)

**Comment by SwayamInSync** [comment]
All the known issues along with this one are resolved in the latest version of this repository.
Closing this issue here.

Thanks
